peps/pep-0807.rst

```diff
@@ -0,0 +1,398 @@
+
+Rejected Ideas
+==============
+
+"Lateral" Discovery
+-------------------
+
+This PEP's discovery mechanism uses the ``.well-known`` location scheme
+defined in :rfc:`8615`. This scheme is widely adopted by machine-to-machine
+protocols, including OpenID Connect itself
+(for `OpenID Connect Discovery <https://openid.net/specs/openid-connect-discovery-1_0.html>`_).
+
+An alternative idea considered was to use a "lateral" discovery mechanism,
+in which the uploading client would attempt discovery by constructing a
+adjacent path relative to the upload URL. For example, for
+``https://upload.example.com/legacy/``, the uploading client would
+attempt to discover Trusted Publishing support at
+``https://upload.example.com/legacy/pytp`` (or some equivalent).
+
+The advantage of this approach is that it doesn't require index operators
+to have control over their (sub-)domain, which the ``.well-known`` scheme
+expects (as well-known URIs can only be served from the root of a domain).
+
+However, this approach also has downsides:
+
+* It assumes that arbitrary indices can provide an adjacent path without
+  interfering with existing functionality, which isn't necessarily true.
+  For example, a given third-party implementation may already use
+  all routes under ``/legacy/{*}`` for other purposes.
+* It's less consistent with existing machine-to-machine protocol
+  conventions, which overwhelmingly use the ``.well-known`` scheme. Developing
+  a custom location scheme here would require additional informational
+  materials for server administrators and operators who are accustomed
+  to the ``.well-known`` scheme.
+
+"Implicit" Discovery
+--------------------
+
+Another alternative idea considered was the perform "implicit" discovery,
+similar to what PyPI currently does for Trusted Publishing: instead of an
+explicit :ref:`discovery <discovery>` step, the uploading client could jump
+straight to attempting the audience and token minting steps, and
+handle any errors that arise.
+
+The advantage of this approach is simplicity: it eliminates the network
+round-trip needed for the discovery step, and eliminates the indirection
+of obtaining the audience and mint token endpoints from the discovery
+response.
+
+This approach too has downsides:
+
+* It implicitly limits a given domain to a single index/upload implementation,
+  since the implicit "discovery" step on PyPI is to construct the audience
+  and mint token endpoints against the base domain of the upload URL.
+  This limitation is acceptable in the context of a single index host
+  like PyPI, but does not generalize to other index topologies (like
+  index hosts that provide isolated private indices).
+* It relies on entirely static endpoint construction rules for
+  the audience and mint token endpoints, which means significant disruption
+  to existing clients if those endpoints ever need to change.
+
+.. rubric:: Footnotes
```

Comment:
We use a full section for footnotes:
```suggestion

Footnotes
=========
```

Reply:
Thanks -- I copied the style in a few other PEPs, but it felt wrong to not have a section for them!